### ISSUE TITLE
linearham 2.0: the Awakening

### ISSUE THREAD

**Issue by dunleavy005** [matsengrp/linearham#44]
As we discussed last week, a linearham re-structuring is required if we want to have objects that can be easily used for HMM hidden state sampling.

Remember for a standard HMM with `x_1`, ..., `x_n` being the hidden states and `y_1`, ..., `y_n` being the observed states (Scott, 2002):

![gif latex](https://user-images.githubusercontent.com/18724296/41993910-964c0daa-7a01-11e8-97a8-2bc8681f86ed.gif)

where the "alphas" are the cached HMM forward probabilities.  This suggests a forward pass through the model to cache the "alphas", and then a backward pass to sample the hidden states. To make this easier to implement, we need to have data structures that store the quantities shown in the formulas above.

Currently, we've set things up (using Smooshables) that allows for easy computation of the HMM likelihood, but impossible hidden state sampling.  One way to see this: we "smoosh" V, D, J marginal probability matrices left-to-right and then add these "Chains" up in our "Pile" top-to-bottom, when we really need to marginalize probabilities top-to-bottom at each site position as we go left-to-right (i.e. order of operations is swapped).

**I propose the following:**

Partition the HMM into distinct regions, the "germline" regions and the "junction" regions.  The "germline" regions are guaranteed to contain germline states (known via the input S-W flexbounds) and thus have collapsed state spaces.  The V-D and D-J "junction" regions are trickier to handle because there are many possible state configurations.

Our test example:
![img_20180720_083850](https://user-images.githubusercontent.com/18724296/43012772-e896684c-8bfb-11e8-851e-9080c506d8f4.jpg)

In the test example above, the V "germline" region would extend from sites 0-3 and the state space would consist of the different V germline genes (i.e. IGHV1-2*56, ....) with a cached forward probability per state.  This would be computed by doing a "Smooshable-like" calculation with no exit probability (`P(pad) P(emit 1) P(1 -> 2) P(emit 2) P(2 -> 3) P(emit 3)` for our example V gene).  We also have a transition probability matrix from the V "germline" region to the V-D "junction" region at sites 4-7.  The state space here is (`"V,3", "V,4", "D, N_a", "D, N_c", "D, N_g", "D, N_t", "D, 0", "D, 1", "D, 2"`), thus the t.p.m. is (1 x 9) in our example.  The forward probability matrix for this "junction" region would have rows indexed by the different states and columns by the different sites (i.e. `4, 5, 6, 7`).  There would then be a transition probability matrix from the V-D "junction" region to the D "germline" region, which would be treated similarly to the V "germline" region.  The remaining regions have a similar structure.

Computationally, we'd store all the ^ matrices in some class (Data?).  Of course, we want to compute the minimal number of libpll likelihoods for emissions so utilizing the work in #43 will be important.  One big difference is that I think our initial thoughts of using a `{germline base, germline rate, MSA pos.} -> xMSA pos.` map should be adjusted to `{germline base, MSA pos.} -> xMSA pos.` as libpll doesn't seem to be adding per-site rates in the near future (let's stick to 4-rate gamma?).  Extracting emission likelihoods is simple as each HMM state encodes a germline base and we iterate across the MSA positions.  To make our lives easier, we could even store a `vector<string>` holding the states in a particular region and a `vector<int>` holding the corresponding germline bases.  Right now, we have two maps `germ_xmsa_indices_` and `nti_xmsa_indices_` that stores the xMSA indices used to extract emission probs. in germline/NTI regions, respectively; we could do something analogous for the "germline" and "junction" regions in this proposal.

Given all of the storage requirements described ^, we'd have to use the parsed partis HMM parameter information (from the YAML files) to "fill in" these matrices/vectors/maps.  Once we do that, we do our forward pass to fill in the forward probability entries and then can sample hidden states according to the backward recursion.

It does feel a bit like re-inventing the wheel WRT to partis's HMM state structure, but there seem to be a few benefits of having linearham:
1) reduced state space in the "germline" regions, leading to faster HMM likelihood computation (otherwise we're doing "quadraticham" and would be slow when doing weighted SRS for tons of trees),
2) hidden state sampling allows us to characterize naive sequence uncertainty (unlike partis viterbi),
3) linearham can work on single sequence reads (star-tree) or phylogenetic MSAs (phylo-tree).

Thoughts, @matsen ??

**Comment by dunleavy005** [comment]
1. Yes, definitely, I'll try to do this sometime this upcoming week.
2. Discretization to 4 rates (using mean in each discrete category, which is default in libpll). Sorry for not being clear.
3. Yes, we try all "germline bases" in the junction area, which I agree is not correct wording. I'll change that.

I was able to implement the forward algorithm using the structure stated above and the resulting log-likelihood checks out with our Smooshable log-likelihood on the test example!  However, I did realize I was making a slightly different flexbounds assumption about "germline" vs "NTI" regions than was made using the Smooshable formulation (at first, the log-likelihoods didn't equate, but when I made the assumptions match, they were identical).

Looking at the test picture above, the V-D "junction" region goes from `v_r[0]` to `d_l[last]` and I allow any base, NTI or other possible germline to be the hidden states.  In particular, in my formulation, `D, 0` and `D, 1` are allowed hidden states, which is unlike the Smooshable setup allowing only `D, 2` and `D, 3` to be possible D gene landing-in points.  I think this difference amounts to how much we rely on S-W to collapse the hidden state space.  My proposal only uses S-W to determine what site positions are guaranteed to be "germline" (i.e. "germline" regions) and lets the rest be germline or NTI (i.e. "junction" regions), whereas Smooshables additionally restrict the possible landing-out/landing-in positions for germline genes in the junction.  The way I've coded things up, it's just a matter of adding more state spaces, more caching of transition matrices, etc. etc. to get my proposal to the Smooshable assumption, but given we know these "junction" regions are super short compared to "germline" regions, the speed increases may not be substantial WRT to making the codebase even more complicated.

Maybe we can chat about this more IRL?
